Curl: Use `typed: strict`

samford opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This upgrades `utils/curl.rb` to `typed: strict`, which requires a number of changes to pass `brew typecheck`. The most straightforward are adding type signatures to methods, adding type annotations (e.g., `T.let`) to variables that need them, and ensuring that methods always use the expected return type.

I had to refactor areas where we call a `Utils::Curl` method and use array destructuring on a `SystemCommand::Result` return value (e.g., `output, errors, status = curl_output(...)`), as Sorbet doesn't understand implicit array conversion. As suggested by Markus, I've switched these areas to use `#stdout`, `#stderr`, and `#status`. This requires the use of an intermediate variable (`result`) in some cases but this was a fairly straightforward substitution.

I also had to refactor how `Cask::URL::BlockDSL::PageWithURL` works. Currently it uses `page.extend PageWithURL` to add a `url` attribute but this reworks it to subclass ~`String`~ `SimpleDelegator` and use an `initialize` method instead. This achieves the same goal but in a way that Sorbet can understand.

-----

Besides that, this adds more tests to `curl_spec.rb` to increase test coverage. This brings almost all of the methods that don't make network requests up to 100% line and branch coverage (the exception being some guards in `parse_curl_output` that shouldn't happen under normal circumstances). Overall coverage changes are as follows:

| `brew tests` Command       | Line % | Branch % |
|----------------------------|--------|----------|
| `--only utils/curl` before | 47.55  | 32.20    |
| `--only utils/curl` after  | 55.24  | 40.23    |
| `--online` before          | 64.91  | 47.46    |
| `--online` after           | 67.83  | 53.45    |

In the process of writing more tests for `parse_curl_response`, I made some tweaks to remove checks for conditions that shouldn't ever be true (e.g., `match["code"]` isn't optional, so it will be present if `HTTP_STATUS_LINE_REGEX` matches) and to refactor some others. I contributed this method a while back (9171eb2), so this is me coming back to clarify some behavior.

-----

Though this is worthwhile in its own right (#17297), this is a prelude to some potential `curl_args` changes to unblock a livecheck feature that I've been working on for some time.

dduugg commented on Library/Homebrew/utils/curl.rb:
```diff
@@ -140,9 +146,23 @@ def curl_args(
 
+    sig {
+      params(
+        args:              T.untyped,
```
I think these could be `String`, maybe `T.any(Pathname, String)` (or `T.any(Pathname, String, URI::Generic)`, etc.), depending how much work you want to do to update call sites.

MikeMcQuaid replied:
Yeh, we should definitely avoid `T.untyped` here.

issyl0 commented:
Thanks for this! I have a WIP branch for making all of `utils` strictly typed, but I’d not made it to this one yet!

reitermarkus commented on Library/Homebrew/github_packages.rb:
```diff
@@ -169,7 +169,7 @@ def schema_uri(basename, uris)
     # We currently use features newer than the last one (v1.0.2).
     url = "https://raw.githubusercontent.com/opencontainers/image-spec/170393e57ed656f7f81c3070bfa8c3346eaa0a5a/schema/#{basename}.json"
-    out, = Utils::Curl.curl_output(url)
+    out, = Utils::Curl.curl_output(url).to_a
```
```suggestion
    out = Utils::Curl.curl_output(url).stdout
```

reitermarkus commented on Library/Homebrew/system_command.rb:
```diff
@@ -427,6 +427,7 @@ def success?
     def to_ary
       [stdout, stderr, status]
     end
+    alias to_a to_ary
```
Probably better to use `#stdout`, `#stderr` or `#status` instead of explicitly calling `#to_a`.

samford replied:
The only downside is that using the explicit methods could potentially require up to four lines if using all of the values in a way that requires them to be stored in variables. However, it doesn't seem like that will be a common issue in practice, so I'm good with this approach.

For example, the `curl_output` call in `Utils::GitHub::API.open_rest` uses `output`, `errors`, and `status`, so we would need to store the return value in an intermediate variable (e.g., `result`) to be able to use all of the values. If we replace `errors` with `result.stderr` and `status.success?` with `result.status.success?`, we end up using the same number of lines.

I'll address this in the next push as well 👍

MikeMcQuaid commented on Library/Homebrew/utils/curl.rb:
```diff
@@ -140,9 +146,23 @@ def curl_args(
+        env:               T::Hash[String, String],
+        timeout:           T.nilable(T.any(Integer, Float)),
+        use_homebrew_curl: T::Boolean,
+        options:           T.untyped,
```
Let's type this, too. Ideally: we'd be replacing `options` usage with more actual keyword parameters instead in all cases. For now at least: making this a hash with a typed key would be better.

reitermarkus replied:
It's already implicitly a hash by being a splat parameter, `T.untyped` only refers to the values in this case.

samford replied:
This is something that tripped me up in the past until I read the [Sorbet documentation for rest parameters](https://sorbet.org/docs/sigs#rest-parameters). As Markus mentioned, the type for splat parameters in signatures is the type of the hash _values_, not the hash itself. It's a weird idea but if we use `options: T::Hash[Symbol, T.untyped]` here it will produce type errors like:

```
download_strategy.rb:593: Expected T::Hash[Symbol, T.untyped] but found TrueClass for argument options https://srb.help/7002
     593 |    curl_download resolved_url, to:, try_partial: @try_partial, timeout:, use_homebrew_curl: true
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Expected T::Hash[Symbol, T.untyped] for argument options of method Utils::Curl#curl_download:
    utils/curl.rb:231:
     231 |        options:     T::Hash[Symbol, T.untyped],
                  ^^^^^^^
  Got TrueClass originating from:
    download_strategy.rb:593:
     593 |    curl_download resolved_url, to:, try_partial: @try_partial, timeout:, use_homebrew_curl: true
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

-----

> I think it'd be nicer to have explicit options here (eventually) rather than the implicit splat hash

If I'm remembering correctly, `options` covers method parameters and/or arguments to `curl_args`. This PR is a precursor to some work to improve how we're handling `curl_args` arguments in other methods (e.g., `curl_with_workarounds`, so that will cover the latter if I'm able to get it to work how I want. If that works, then it should be easier at that point to add explicit parameters instead of `options` (i.e., something to keep in mind for later).

MikeMcQuaid replied:
Fine to handle later 👍🏻

MikeMcQuaid reviewed (commented):
Thanks @samford! Looking good so far.

samford commented:
Changes in the latest push:

* I switched the `args`/`extra_args` type from `T.untyped` to `String`, per @dduugg's comment. We could use `T.any(Pathname, String)` but there was only one pathname that I've seen so far, so I added a `#to_s` call instead. Similarly, there were a few areas where a `URI` object is used, so I added `#to_s` calls to those as well (to bring them in line with existing call sites where we're doing the same for other `URI`s). These are curl CLI arguments, so there's maybe something to be said for simply ensuring that these values are strings instead of permitting `Pathname` or `URI::Generic`.
* I refactored areas where we call a `Utils::Curl` method and use array destructuring on a `SystemCommand::Result` return value (e.g., `output, errors, status = curl_output(...)`) to use `#stdout`, `#stderr`, and `#status` instead of my previous `#to_a` approach, per @reitermarkus's comments. I was curious if this would lead to increased verbosity but it ended up being a fairly straightforward substitution (i.e., pretty much the same number of lines).

This passes `brew typecheck` and `brew tests --online` but there may be issues that are only surfaced at runtime, so it would probably be a good idea for folks to do some manual testing with commands that interact with curl before merging.

-----

> Thanks for this! I have a WIP branch for making all of utils strictly typed, but I’d not made it to this one yet!

I looked back at some old Git stashes and apparently I've been intermittently working on this since at least 2022 (there were some challenges along the way, to say the least 😆). I finally finished it late last year and I had been meaning to create a PR but hadn't gotten around to it yet. I saw one of your Sorbet PRs the other day and thought, "Oh, I bet Issy's getting close to `curl`, I better get this out the door." Anyway, I'm glad I managed to save you the trouble!

dduugg commented on Library/Homebrew/cask/url.rb:
```diff
@@ -99,7 +99,7 @@ def initialize(
     class BlockDSL
       # Allow accessing the URL associated with page contents.
-      module PageWithURL
+      class PageWithURL < String
```
I've had bad experiences subclassing the core library, as have [others](https://steveklabnik.com/writing/beware-subclassing-ruby-core-classes/). A more common rubyism is to subclass [`SimpleDelegator`](https://docs.ruby-lang.org/en/3.3/SimpleDelegator.html), then initiate it with the string you want to delegate to.

dduugg replied:
Looks great! I wasn't sure how sorbet-compatible it would be, but you were able to make it work without any rbi shenanigans. 🅰️➕ work!

MikeMcQuaid reviewed (approved):
Should be good once @dduugg's comment addressed!

samford commented:
Thanks everyone for all the helpful reviews! Merging this now, as I'll be around to address any potential breakages.
